source/common/formatter/substitution_formatter.cc

```diff
@@ -1918,6 +1925,69 @@ UpstreamHostMetadataFormatter::UpstreamHostMetadataFormatter(const std::string&
                           return host->metadata().get();
                         }) {}
 
+ProxyProtocolTlvsFormatter::ProxyProtocolTlvsFormatter(const std::string& tlv_type) {
+  ASSERT(!tlv_type.empty());
+  // Specified tlv_type must be parsable as an int.
+  try {
+    tlv_type_ = std::stoi(tlv_type);
+  } catch (...) {
+    ENVOY_LOG_MISC(error, "Invalid parameter provided for PROXY_PROTOCOL_TLVS header: {}", tlv_type);
+    // Arbitrary int that won't be used for proxy procotol TLV type. 
+    tlv_type_ = -1; 
```

Comment:
[q] I think this should be safe as an invalid value flag because the [proxy protocol RFC](https://www.haproxy.org/download/1.8/doc/proxy-protocol.txt) defines the following values as registered TLV types: 
```

The following types have already been registered for the <type> field :

        #define PP2_TYPE_ALPN           0x01
        #define PP2_TYPE_AUTHORITY      0x02
        #define PP2_TYPE_CRC32C         0x03
        #define PP2_TYPE_NOOP           0x04
        #define PP2_TYPE_UNIQUE_ID      0x05
        #define PP2_TYPE_SSL            0x20
        #define PP2_SUBTYPE_SSL_VERSION 0x21
        #define PP2_SUBTYPE_SSL_CN      0x22
        #define PP2_SUBTYPE_SSL_CIPHER  0x23
        #define PP2_SUBTYPE_SSL_SIG_ALG 0x24
        #define PP2_SUBTYPE_SSL_KEY_ALG 0x25
        #define PP2_TYPE_NETNS          0x30
```

Please let me know if you have any concerns with this approach!

Reply:
The only example I see in the codebase where we validate a TLV type is in `proxy_protocol.pb.validate.cc`, which is generated at compile-time by `protoc-gen-validate` from [this proto file](https://github.com/10gen/envoy-serverless/blob/pass-proxy-protocol-tlv-in-http-headers/api/envoy/extensions/filters/listener/proxy_protocol/v3/proxy_protocol.proto). This is what I see in the generated code: 
```
bool Validate(const ::envoy::extensions::filters::listener::proxy_protocol::v3::ProxyProtocol_Rule& m, pgv::ValidationMsg* err) {
...
 if (m.tlv_type() >= 256) {
	{
    std::ostringstream msg("invalid ");
    msg << "RuleValidationError" << "." << "TlvType";
    msg << ": " << "value must be less than 256";
    *err = msg.str();
    return false;
    }
}
...
} 
The above block is generated by this line in the `proto` file:
```
    uint32 tlv_type = 1 [(validate.rules).uint32 = {lt: 256}];

```
We could do the same thing here and check if it's less than 256 in the constructor after `stoi`

Reply:
Checking it's less than 256 makes sense to me. Failing early sounds good to me. Is that the pattern in this file for other formatters?

Reply:
The pattern varies based on the type of validation taking place. Missing param values or too many params make Envoy startup fail ([example](https://github.com/10gen/envoy-serverless/blob/aa484fbfcfd6aa19544a860d45e9794ff0f2ee73/source/common/formatter/substitution_formatter.cc#L2128-L2136)), same with basic formatting issues with params passed in ([example](https://github.com/10gen/envoy-serverless/blob/aa484fbfcfd6aa19544a860d45e9794ff0f2ee73/source/common/formatter/substitution_formatter.cc#L339)). 

When it comes to validating the actual values passed in as params, behavior varies. 
[Here](https://github.com/10gen/envoy-serverless/blob/aa484fbfcfd6aa19544a860d45e9794ff0f2ee73/source/common/formatter/substitution_formatter.cc#L1937-L1938), we fail startup if a value other than `PLAIN` or `TYPED` is passed in. 
[Here](https://github.com/10gen/envoy-serverless/blob/aa484fbfcfd6aa19544a860d45e9794ff0f2ee73/source/common/router/header_formatter.cc#L152-L156), Envoy startup succeeds and simply doesn't add an HTTP header if the provided value type of the metadata is unsupported. 

I agree that it sounds best to fail fast for this case. It seems more clear than allowing Envoy startup to succeed and not adding a header. This would help decrease "user error" cases where the `tlv_type` is just incorrectly entered into the envoy config.